src/Symfony/Component/DependencyInjection/Compiler/CheckDefinitionValidityPass.php

```diff
@@ -43,7 +43,7 @@ public function process(ContainerBuilder $container)
             }
 
             // non-synthetic, non-abstract service has class
-            if (!$definition->isAbstract() && !$definition->isSynthetic() && !$definition->getClass()) {
+            if (!$definition->isAbstract() && !$definition->isSynthetic() && !$definition->getClass() && (!$definition->getFactory() || '.' !== substr($id, 0, 1))) {
```

Comment:
Any reason to use `substr()` instead of `$id[0]`?

Reply:
updated to a stricter regexp actually